Add tutorial for driven solver (adaptive and uniform) (#702)
awslabs/palace: docs/src/examples/tutorial_driven_uniform_v_adaptive.md

    This tutorial is advanced. It goes into more of the algorithmic detail, to help users effectively use the driven solvers. However, some of these algorithmic choices are considered “internal” to
    *Palace*. They may change at any time if the developers decide that alternative approaches are better.

    Please proceed with caution.

!!! note

    For the CPW example, the data can be generated with the script `examples/cpw/cpw_tutorial_lumped_driven.jl` and the result plots generated with `examples/cpw/cpw_tutorial_lumped_driven_plots.py`.

    For the transmon example, the data can be generated with the script `examples/transmon/transmon_tutorial_driven.jl` and the result plots generated with `examples/transmon/transmon_tutorial_driven_plots.py`.

## Driven Solver Quick-Start

*Palace* has two modes for frequency-domain driven simulations: uniform and adaptive. The uniform
solver runs one independent FEM solve per output frequency, which is reliable but slow. The adaptive
solver runs a small number of FEM solves and constructs a reduced-order model (ROM) that
interpolates the response at all other output frequencies. This is fast, but requires more careful

> set-up and validation.

Reviewer: ```suggestion
interpolates the response at all other output frequencies. This is fast, but introduces a source of error compared to the full model evaluation, which has to be controlled.
```

the uniform solver. Any `"AdaptiveTol"` > 0 calls the adaptive solver.

```json
"Driven": {
  "Samples": [ {"Type": "Linear", "MinFreq": 3.5, "MaxFreq": 7.0, "FreqStep": 0.1} ],
  "AdaptiveTol": 1e-3 // Use adaptive solver
}
```

The `"Samples"` specification defines the output frequency grid for both solvers. For the adaptive
solver, this grid can be very fine with only a small additional cost per point. All output files
(`domain-E.csv`, `port-S.csv`, etc.) have the same format in both cases.

When studying a new model, users should validate the adaptive solver against the uniform solver on a
coarse output `"Samples"` grid and at key frequencies. This helps build intuition about the response
and the approximation error of the adaptive solver for the specific quantities of interest. Users
should tighten tolerances and re-validate as needed before running at fine frequency resolution.

The following are important tuning configurations:

  - [`"AdaptiveTol"`](../config/solver.md#solver%5B%22Driven%22%5D): start with `1e-3` on a coarse
    grid for S-parameter sweeps. Tighten this tolerance as needed during the validation against the
    uniform solver or in production runs.

>   - `"Linear"/"Tol"`: set the linear solver tolerance to be substantially smaller than

Reviewer 2: We usually use the notation `["Linear"]["Tol"]`  for nested configs. To be more precise, the other doc pages often refer to configs arguments as `config["Solver"]["Linear"]["Tol"]`.

>     `"AdaptiveTol"`. The adaptive solver is based on an interpolation and can be unusually sensitive

Reviewer: ```suggestion
    `"AdaptiveTol"`. The adaptive solver is based on an interpolation and can be very sensitive
```

    to errors in the linear solver. Increase this tolerance in your validation runs and ensure the

>     output is no different. Should *Palace* log warnings like `Minimal rational interpolation encountered rank-deficient matrix` try tightening this tolerance substantially.

Reviewer 2: Is this something we should explicitly say in the log message in Palace?

  - [`"AdaptiveMaxSamples"`](../config/solver.md#solver%5B%22Driven%22%5D): cap on full linear
    solves per excitation (default: 20). If the log indicates that you exceed this bound, it means
    the adaptive solver has not reached the target tolerance and this bound should be increased. If
    the adaptive solver never converges, it might indicate a numerical problem.
  - [`"AdaptiveConvergenceMemory"`](../config/solver.md#solver%5B%22Driven%22%5D): number of
    consecutive "safety" samples that have to be below the tolerance threshold before convergence.
    Defaults to 2, which is small but fast. Increase this number to 3+ if you see early termination
    without convergence.

The rest of this tutorial will illustrate the driven solvers on two different examples (CPWs and

> transmon), explain the adaptive algorithm in detail, and give more detailed user guidance.

Reviewer: ```suggestion
The rest of this tutorial will illustrate the driven solvers on two different examples (CPWs and
and the single transmon), explain the adaptive algorithm in detail, and give more detailed user guidance.
```

## Revisiting the Two Coplanar Waveguide Example (Part I)

We start by re-visiting the driven response of the [two CPW example](cpw.md), using the mesh
[`mesh/cpw_lumped_0.msh`](https://github.com/awslabs/palace/blob/main/examples/cpw/mesh/cpw_lumped_0.msh)
that places lumped ports at the end of the CPWs.

### Uniform Solver

First, we set up a uniform driven simulation using the configuration file

> [`cpw_lumped_uniform_convergence.json`](https://github.com/awslabs/palace/blob/main/examples/cpw/cpw_lumped_uniform_convergence.json).

Reviewer 2: This file was not included in the commit

In this example, we will only be driving the single lumped port with `"Index": 1`. The `Solver`
section of the configuration is

```json
"Solver" : {
  "Order" : 2,
  "Device": "CPU",
  "Driven": { "Samples": [ {"Type": "Linear", "MinFreq": 2.0, "MaxFreq": 32.0, "FreqStep": 0.1} ] },
  "Linear": {"Type": "Default", "Tol": 1.0e-12, "MaxIts": 1000, "ComplexCoarseSolve": true}
}
```

*Palace* calls the uniform driven solver when
[`"AdaptiveTol"`](../config/solver.md#solver%5B%22Driven%22%5D) in `Solver/Driven` is `0.0` or not
specified. The uniform solver will iterate over all sample points specified in
`Solver/Driven/Samples` and independently solve the linear equation for the electric field at the
sample frequency point ``f``. We will also refer to these solves as "full" or "High-Dimensional
Solves" (HDM) since they solve the full linear problem on the finite element space. Here we are
sampling a linear grid between ``2~\mathrm{GHz}`` and ``32~\mathrm{GHz}`` in steps of ``\Delta f = 0.1~\mathrm{GHz}``, but [users can provide more complicated sample

> specification](../config/solver.md#solver%5B%22Driven%22%5D%5B%22Samples%22%5D).

Reviewer 2: ```suggestion
specifications](../config/solver.md#solver%5B%22Driven%22%5D%5B%22Samples%22%5D).
```

The linear solver tolerance `"Tol": 1.0e-12` is chosen to be very small. Such a small tolerance is
near the limit of what *Palace*'s solvers can reach using double precision. The error of the uniform
linear solve ``\bm{A}(\omega) \bm{x} = \bm{b}(\omega)`` also depends on the condition number of the
system matrix ``\bm{A}(\omega)`` at each frequency ``\omega = 2 \pi f``. Importantly, if the system

> has resonances (poles) near the real axis, the linear solve may become very poorly conditioned. This

Reviewer 2: ```suggestion
has resonances (poles) near the real axis, the linear solve may become poorly conditioned. This
```

simulation, *Palace* will only compute the column ``S_{i1}``. We plot the magnitude below.

```@raw html
<br/><p align="center">
  <img src="../../assets/examples/driven_ua_cpw_domain_sparam_uniform.svg" width="80%" />
</p><br/>
```

This data is equivalent to the data previously shown in the [CPW example](cpw.md), since we consider
the same physical model although we have used slightly different solver parameters as shown above.

### Adaptive Solver

The primary downside of the uniform driven solver is its slow performance. Every output sample has
to be individually calculated, which becomes prohibitive on a fine output grid. This is the problem
the adaptive solver addresses.

To perform a driven simulation with the adaptive driven solver, we set
[`"AdaptiveTol"`](../config/solver.md#solver%5B%22Driven%22%5D) to `1e-1` in the configuration. We
will define this quantity more precisely later. Simulation output, like `domain-E.csv`, has the same
structure as the uniform solver — they are evaluated on the same frequency grid specified by
`"Samples"` in the configuration file. However, the adaptive solver constructs a different

> "internal" frequency grid on which it does HDM solves. This internal grid is automatically chosen by

Reviewer: ```suggestion
*internal* frequency grid on which it does HDM solves. This internal grid is automatically chosen by
```
generally the same thing for most of the `""` of terms here, they're not qualified so much as emphasized.

The idea of the adaptive solver is to transform this problem with a large dimension ``N`` into a
linear problem with much smaller dimension ``n`` that approximates the high-dimensional model well.

```math
\left[\bm{K}_r + i\omega \bm{C}_r - \omega^2 \bm{M}_r \right] \bm{x}_r = i \omega \bm{b}_r
```

This also requires a method to efficiently recover the high-dimensional field solution ``\bm{x}``
from the reduced ``\bm{x}_r``. If we have found such a reduced model, we can quickly solve the small
linear system at a specific ``\omega`` and recover the full solution and all output measurements. We
refer to the procedure of finding the model as the “offline” or “training” phase, and to using it to
calculate many output samples as the “online” phase.

There is a large literature on constructing dimensional reductions of linear systems and we refer to

> the literature for an introduction [1,2]. *Palace* finds a set of ``n`` real orthogonal vectors

Reviewer: Link the anchors.

``\bm{Q}`` that capture the full response of the system in the frequency range of interest
``[f_\mathrm{min}, f_\mathrm{max}]``. The projection ``\bm{K}_r = \bm{Q}^T \bm{K} \bm{Q}``,
``\bm{b}_r = \bm{Q}^T \bm{b}``, etc., and the prolongation ``\bm{x} = \bm{Q} \bm{x}_r`` are
particularly simple.

The set of basis vectors ``\bm{Q}`` that *Palace* uses are the orthogonalized components of the HDM

> solutions ``\bm{x}^*`` at the "internal" sampling frequencies ``f_\mathrm{sample} \in [f_\mathrm{min}, f_\mathrm{max}]``. In fact, *Palace* uses ``\mathrm{Re}\bm{x}^*``,

Reviewer: Re and Im probably need `()` or `{}`, your call.

``\mathrm{Im}\bm{x}^*`` as separate vectors so that the basis ``\bm{Q}`` is real. As we add more
sampling frequencies, the basis of the reduced order model grows and increases the accuracy of the
projection, until we satisfy a convergence criterion. This construction is essentially a type of
rational interpolation on the whole domain. It is also related to the mathematics of rational Krylov
spaces [2].

The Gram-Schmidt orthogonalization algorithm used to construct ``\bm{Q}`` from the HDM solutions is
determined by the user option
[`"AdaptiveGSOrthogonalization"`](../config/solver.md#solver%5B%22Driven%22%5D), although a user
should generally not need to change this. By default, it uses classical Gram-Schmidt with
reorthogonalization which is a good high-precision option.

### Choosing Internal Sample Frequencies and Convergence

The above construction does not tell us how to efficiently choose the location of internal sample

> frequencies ``f_\mathrm{sample}``. For this we use the greedy sampling algorithm developed by

Reviewer: ```suggestion
The above construction does not explain how to efficiently choose the location of internal sample
frequencies ``f_\mathrm{sample}``. For this we use the greedy sampling algorithm developed by
```

Pradovera in [3]. This is based on a type of rational (barycentric) interpolation for the solution
``\bm{u}`` of linear systems in frequency

```math
\bm{u}(\omega) = \frac{\sum_i w_i \bm{u}(\omega_i) / (\omega - \omega_i)}{\sum_i w_i / (\omega -
\omega_i)},
```

where the ``\omega_i`` are sample frequencies, ``\bm{u}(\omega_i)`` solutions at those frequencies,
and ``w_i`` some weight coefficients fitted. To use this interpolation for our driven
electromagnetic solver, we linearize the quadratic ``KCM`` operator as a linear operator and
identify ``\bm{u}^T = (\bm{x}^T, i \omega \bm{x}^T)``. This type of linearization is familiar to

> physicists as going from Newton's equations (second order in time) to Hamilton's equations (first

Reviewer: ```suggestion
identify ``\bm{u}^T = (\bm{x}^T, i \omega \bm{x}^T)``. This type of linearization may be familiar to
physicists as going from Newton's equations (second order in time) to Hamilton's equations (first
```

solves. *Palace* prints the location of the sampling points found during the adaptive solver
simulations to the log-file.

Convergence of the procedure is based on the relative error of the HDM electric field solution
``\bm{x}_\mathrm{HDM}`` compared to the solution calculated by the ROM:

```math
\varepsilon = \frac{\vert\vert \bm{x}_\mathrm{HDM}(f^*) -
\bm{x}_\mathrm{ROM}(f^*)\vert\vert}{\vert\vert \bm{x}_\mathrm{HDM}(f^*) \vert\vert}
```

Specifically, we declare convergence when this error is below the user-defined tolerance
``\varepsilon < \varepsilon_\mathrm{tol}``, set by
[`"AdaptiveTol"`](../config/solver.md#solver%5B%22Driven%22%5D), for a number of consecutive
frequency samples, set by
[`"AdaptiveConvergenceMemory"`](../config/solver.md#solver%5B%22Driven%22%5D). Note that the norm
``\vert\vert \bm{x} \vert\vert`` is the L2 norm in the finite element space. So this is similar to,
but not the same as, the physical norm computed from the overlap in space ``\int \bm{E}^*(\bm{r}) \cdot \bm{E}(\bm{r}) d^3r``, which corresponds to a mass-weighting in finite-elements
``\bm{x}^T\bm{M}\bm{x}``. This difference between L2 and mass-weighted overlap can be significant on

> highly anisotropic meshes.

Reviewer: This seems like a good moment to mention that the indicator we use is an estimator, not a strict bound, so the memory is a way to mitigate against falsely ending when the true error is still large. A reader who doesn't catch that detail might be a bit confused why we don't trust the first sample set below the tolerance level. 

Edit: I see you bring this up in the next section, I was just getting ahead.

As mentioned above, the relative error is of the electric fields over the global domain. However,
the relative error on a derived measurement need not satisfy that error bound. This occurs when the
measurement is difficult, such as integrating the electric field in a region where the field is
tiny.

### Adaptive Solver Problems and User Guidance

The rational interpolation approach above is extremely efficient at building a basis using only a
very small number of HDM solves. However, there are two challenges that we encounter in practical
use.

> **Early termination** due to accidental convergence. As discussed above, the convergence criterion

Reviewer: Looking at the example rendering of this at https://awslabs.github.io/palace/previews/PR702/examples/tutorial_driven_uniform_v_adaptive/, the "**" "**" isn't coming out. It might be worth fiddling the format a little to identify the breaks here, as with the paragraph break I was momentarily confused why "To avoid this" was the second point.

threshold ``\varepsilon_\mathrm{tol}``. Although the algorithm picks ``f^*`` to be the position
where the error indicator is largest, the error evaluated there can sometimes be smaller than the
error expected from the interpolation. This means that the solution looks more converged than it is
and the algorithm terminates early. This is discussed in the paper by Pradovera [3].

To avoid this, we introduce a memory. The algorithm only terminates once a certain number of
consecutive samples in a row are below tolerance. This number is defined by the
[`"AdaptiveConvergenceMemory"`](../config/solver.md#solver%5B%22Driven%22%5D) configuration. The
default is `2`, which is small. Increasing this number to say `3` or `4`, especially in models where
the user does not understand the convergence properties, can be helpful. However, each HDM sample
adds a substantial computational cost to the training phase.

**Numerical problems** due to finite precision. The mathematical formulation of the rational
interpolation for the error indicator assumes that the solution at the sample points is exact.
However, in reality there is the linear solver tolerance

> [`"Linear"/"Tol"`](../config/solver.md#solver%5B%22Linear%22%5D) that sets the error on this and can

Reviewer: Normally we put the links with the `[]`. Can be better fixed once your other PR is merged with the good hyperlinking capability.

warnings. This breaks the next best frequency prediction, which causes the adaptive solver to
frequently oversample certain frequency regions and sometimes undersample other regions. While this
does not break the ROM structure as such — it will accommodate a larger number of basis states — the
quality of that basis can be poor and the desired tolerance not reached.

The number of HDM samples for the ROM is capped by the configuration
[`"AdaptiveMaxSamples"`](../config/solver.md#solver%5B%22Driven%22%5D) (default: 20), after which
the solver just uses the capped ROM to do the computation. That cap is helpful to bound the
oversampling cases due to numerical issues. However, under normal conditions the user should
increase [`"AdaptiveMaxSamples"`](../config/solver.md#solver%5B%22Driven%22%5D) to guarantee
accuracy.

As discussed above, a user can also force certain frequency points to be added to the PROM before
the adaptive sampling starts by using [`"AddToPROM": true`](../config/solver.md#solver%5B%22Driven%22%5D%5B%22Samples%22%5D). This is primarily a
debugging tool. Routine usage is not recommended — adding sample by hand negates the major benefit

> of the adaptive solver to “choose the best” samples for a given accuracy. However, this option might

Reviewer: ```suggestion
of the adaptive solver to efficiently choose samples based on the observed approximation error. However, this option might
```

help investigate convergence issues. If adding points, the user should be careful not to
accidentally set [`"AddToPROM": true`](../config/solver.md#solver%5B%22Driven%22%5D%5B%22Samples%22%5D) on a dense grid, since this

> will perform a very large number of HDM solves and keep them in computer memory for the duration of

Reviewer: ```suggestion
will perform a very large number of HDM solves and keep them in memory for the duration of
```

the relative error can be very bad! However, this should not be surprising. The only scattering
matrix that has a large value (of order 1) is ``S_{21}`` and we see that the relative error there is
well behaved. The bad relative error in the other plots is because the magnitude of ``S_{i1}`` is
tiny to begin with, which magnifies the relative error. For example, the magnitude ``\vert S_{11} \vert`` drops down to around ``-60~\mathrm{dB} = 10^{-3}`` near ``15~\mathrm{GHz}``. That is exactly
the location where the relative error of ``S_{11}`` spikes in the error plot above. The relative
error in ``S_{31}``, ``S_{41}`` is always large since their values are small.

A detailed discussion of error estimates and error propagation from the electric field to derived
quantities is beyond the scope of this tutorial. From a practical perspective, we encourage users to
validate some of the adaptive driven simulations against uniform driven simulations in order to get
a practical sense of the error for different quantities, to help identify an appropriate adaptive

> tolerance and to diagnose convergence errors.

Reviewer: ```suggestion
a practical sense of the error for different quantities, to help identify an appropriate adaptive
tolerance, linear solver tolerance, and to diagnose convergence errors.
```
The interplay here with how accurate the HDM is itself is worth lightly mentioning. Not worth twiddling the adaptive if your HDM samples are all bad to begin with.

``f_\mathrm{max}``. We actually exclude these points from convergence (i.e. they have error
"`inf`"). They are drawn here as points at ``1.0`` only for illustration. As the adaptive algorithm
continues, more sample points are found that decrease the error indicator. In our simulation, we
used the default `"AdaptiveConvergenceMemory"` value of 2. So the simulations converge after two
consecutive points are below `"AdaptiveTol"`. As we can see from the plot above, the error indicator
does not have to decrease monotonically. This means that the error indicator can go above the target
tolerance before dipping below it again. See the discussion above in [Adaptive Solver Problems and
User Guidance](#adaptive-solver-problems-and-user-guidance).

## Driving the Transmon Model

### Set-Up

Let us now apply the driven and adaptive solver to a model of a transmon qubit, which was already
discussed in the [eigenmode tutorial](transmon.md). We assume familiarity with that tutorial. As
discussed there, the model consists of a transmon qubit and a quarter-wave coplanar waveguide
readout resonator coupled to a feedline. There are three lumped ports in this model: ports 1 and 2

> are the ``50~\Omega`` resistive feedline terminations and port 3 is a passive LC element (``L = 14.86\,\textrm{nH}``, ``C = 5.5\,\textrm{fF}``) representing the linearised Josephson junction.

Reviewer 2: ```suggestion
are the ``50~\Omega`` resistive feedline terminations and port 3 is a passive LC element (``L = 14.86\,\mathrm{nH}``, ``C = 5.5\,\mathrm{fF}``) representing the linearised Josephson junction.
```
for consistency with everything else

There are two eigenmodes of particular interest that we discovered in the previous tutorial:

  - A “transmon” mode near ``4.10~\textrm{GHz}`` with ``Q = 1.8 \cdot 10^4``,

>   - A “resonator” mode near ``5.60~\textrm{GHz}`` with ``Q = 7.9 \cdot 10^3``.

Reviewer 2: ```suggestion
  - A “transmon” mode near ``4.10~\marhrm{GHz}`` with ``Q = 1.8 \cdot 10^4``,
  - A “resonator” mode near ``5.60~\mathrm{GHz}`` with ``Q = 7.9 \cdot 10^3``.
```

The eigenmode visualization shows that even the transmon mode has appreciable weight on the readout
resonator and in the feedline.

We will now perform uniform and adaptive driven simulations on this model, by exciting the

> ``50~\Ohm`` resistive ports. We will use the same mesh `examples/transmon/mesh/transmon.msh2` as the

Reviewer 2: ```suggestion
``50~\Omega`` resistive ports. We will use the same mesh `examples/transmon/mesh/transmon.msh2` as the
```

        "Attributes"  : [ 1                                                  ],
        "Permeability": [ 0.99999975,      0.99999975,       0.99999979      ],
        "MaterialAxes": [ [0.8, 0.6, 0.0], [-0.6, 0.8, 0.0], [0.0, 0.0, 1.0] ]
      }
    ]
  },
  "Solver"    : {
    "Driven": {
      "Samples"    : [ {"Type": "Linear", "MinFreq": 3.5, "MaxFreq": 6.5, "FreqStep": 0.025} ],
      "AdaptiveTol": 1e-3
    },
    "Order" : 2,
    "Linear": {"Type": "Default", "Tol": 1.0e-12, "MaxIts": 1000}
  }
}
```

In this case we will be using the multi-excitation feature of *Palace*. Because we have specified
both `"Excitation": 1` on port 1 and `"Excitation": 2` on port 2, *Palace* will iterate over these
excitations separately. This makes no difference to the uniform solver, since it solves every
excitation and frequency sample separately. However, for the adaptive solver, the projective basis
``\bm{Q}`` is shared between all excitations and the convergence criterion ``\varepsilon < \varepsilon_\mathrm{tol}`` is on that combined basis. This means that the output of each excitation
can be different when we simulate them together with one ROM or separately with two ROMs. However,

> these differences are within the adaptive tolerance. See the discussion above.

Reviewer: Do you think it's worth mentioning how this can work well with AMR, as the adaptation is driven by the error in both simulations, and the mesh kept consistent between the two. In the limit of deep AMR convergence it wouldn't matter, but for the ~1e-2 errors we often work at, it can make a big difference to ensure that the same mesh is used for all the excitations. Can possibly put in one of those *info* boxes that @Sbozzolo uses in his examples.

First let us look at the domain energy:

```@raw html
<br/><p align="center">
  <img src="../../assets/examples/driven_ua_transmon_energy_adaptive_sweep.svg" width="80%" />
</p><br/>
```

The conventions are the same as in the plots of the CPW example above. Here, however, the coloured
diamonds mark sample frequencies from the rational interpolations of excitations 1 and 2 together
since they form a combined ROM basis. We see that the adaptive solver converges very quickly. The
error for tolerance `1e-1` is already quite small for all frequencies except near the poles. When we
reach `1e-3` the adaptive solver puts HDM sample points near the pole frequencies and the error
drops dramatically. Tightening the tolerance to `1e-4` and `1e-5` only slightly decreases the error.

The error drop is even more dramatic for the error in the S-parameters:

```@raw html
<br/><p align="center">
  <img src="../../assets/examples/driven_ua_transmon_sparam_adaptive_rms.svg" width="80%" />
</p><br/>
```

In both plots above, the error is more structured than it was in the CPW example. It is worse near
the eigenmode locations and better far away. The adaptive solver also adds HDM solves in a more

> structured manner to “shadow” the eigenmodes. This behaviour is simple to interpret — the response

Reviewer 2: ```suggestion
structured manner to “shadow” the eigenmodes. This behavior is simple to interpret — the response
```
:P

postprocessing without rerunning the simulation. In fact, the ROM matrices ``\bm{K}_r``,
``\bm{C}_r`` and ``\bm{M}_r`` are close in spirit to a circuit. To make that connection precise, we
need to connect the abstract ROM matrix to real electrical signals.

## Literature & References

[1] P. Benner, D. C. Sorensen, and V. Mehrmann, Eds., Dimension Reduction of Large-Scale Systems:
Proceedings of a Workshop held in Oberwolfach, Germany, October 19–25, 2003. in Lecture Notes in
Computational Science and Engineering, no. 45. Berlin, Heidelberg: Springer Berlin Heidelberg, 2005.
doi: 10.1007/3-540-27909-1.

[2] A. C. Antoulas, Approximation of Large-Scale Dynamical Systems. Society for Industrial and
Applied Mathematics, 2005. doi: 10.1137/1.9780898718713.

[3] D. Pradovera, “Toward a certified greedy Loewner framework with minimal sampling,” Adv Comput
Math, vol. 49, no. 6, p. 92, Dec. 2023, doi: 10.1007/s10444-023-10091-7.

[4] D. Pradovera, “Interpolatory rational model order reduction of parametric problems lacking
uniform inf-sup stability,” SIAM J. Numer. Anal., vol. 58, no. 4, pp. 2265–2293, Jan. 2020, doi:

> 10.1137/19M1269695.

Reviewer: Nit: if you can hyperlink to the doi.
